File: app/src/main/java/com/joshualorett/heterogeneousrecyclerviewstudy/lib/ViewHolderCreator.java
package com.joshualorett.heterogeneousrecyclerviewstudy.lib;

import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

/**
 * Creates a view holder.
 * Created by Joshua on 11/20/2016.
 */

public interface ViewHolderCreator {
    /***
     * Get the view holders view type.
     * @return View type
     */
    int getViewType();

    /***
     * Create a new view holder.
     * @param parent {@inheritDoc}
     * @return View holder.
     */
    RecyclerView.ViewHolder createViewHolder(ViewGroup parent);
}

File: app/src/main/java/com/joshualorett/heterogeneousrecyclerviewstudy/lib/ViewHolderCreator.kt
package com.joshualorett.heterogeneousrecyclerviewstudy.lib

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup

/**
 * Create a view holder.
 * Created by Joshua on 11/20/2016.
 */

interface ViewHolderCreator {
    /***
     * Get the view holders view type. This should match at least one ViewHolderBinder's view type.
     * @return View type
     */
    val viewType: Int

    /***
     * Create a new view holder.
     *
     * @param parent {@inheritDoc}
     * @return View holder.
     */
    fun createViewHolder(parent: ViewGroup): RecyclerView.ViewHolder
}
